File: src/main/java/com/hp/jipp/encoding/RangeOfIntegerType.java
package com.hp.jipp.encoding;

import com.google.common.collect.Range;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class RangeOfIntegerType extends AttributeType<Range<Integer>> {
    private final static String TYPE_NAME = "RangeOfInteger";

    static final Attribute.Encoder<Range<Integer>> ENCODER = new Attribute.Encoder<Range<Integer>>(TYPE_NAME) {
        @Override
        public Range<Integer> readValue(DataInputStream in, Tag valueTag) throws IOException {
            Attribute.expectLength(in, 8);
            int low = in.readInt();
            int high = in.readInt();
            return Range.closed(low, high);
        }

        @Override
        public void writeValue(DataOutputStream out, Range<Integer> value) throws IOException {
            out.writeShort(8);
            out.writeInt(value.lowerEndpoint());
            out.writeInt(value.upperEndpoint());
        }

        @Override
        public boolean valid(Tag valueTag) {
            return valueTag == Tag.RangeOfInteger;
        }
    };

    public RangeOfIntegerType(String name) {
        super(ENCODER, Tag.RangeOfInteger, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/RangeOfIntegerType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** Attribute type for [IntRange] attributes */
class RangeOfIntegerType(override val name: String) :
        AttributeType<IntRange>(Encoder, Tag.rangeOfInteger) {

    companion object Encoder : SimpleEncoder<IntRange>("rangeOfInteger") {
        @Throws(IOException::class)
        override fun readValue(input: DataInputStream, valueTag: Tag): IntRange {
            input.takeLength(INT_LEN + INT_LEN)
            val low = input.readInt()
            val high = input.readInt()
            return IntRange(low, high) // vs Closed range
        }

        @Throws(IOException::class)
        override fun writeValue(out: DataOutputStream, value: IntRange) {
            out.writeShort(INT_LEN + INT_LEN)
            out.writeInt(value.first)
            out.writeInt(value.last)
        }
        override fun valid(valueTag: Tag) = valueTag === Tag.rangeOfInteger
    }
}
